spynnaker/pyNN/models/neuron/synaptic_manager.py

```diff
@@ -592,120 +606,575 @@ def _write_synaptic_matrix_and_master_population_table(
+                return False
+            elif (last_slice.hi_atom + 1) != v_slice.lo_atom:
+                return False
+            last_key = key
+            last_slice = v_slice
+        return True
+
+    def __get_app_key_and_mask(self, keys, mask, n_stages, key_space_tracker):
+
+        # Can be merged only if keys are adjacent outside the mask
+        keys = sorted(keys, key=lambda item: item[0])
+        mask_size = KeySpaceTracker.count_trailing_0s(mask)
+        if not self.__check_keys_adjacent(keys, mask_size):
+            return None
+
+        # Get the key as the first key and the mask as the mask that covers
+        # enough keys
+        key = keys[0][0]
+        n_extra_mask_bits = int(math.ceil(math.log(len(keys), 2)))
+        core_mask = (((2 ** n_extra_mask_bits) - 1))
+        new_mask = mask & ~(core_mask << mask_size)
+
+        # Final check because adjacent keys don't mean they all fit under a
+        # single mask
+        if key & new_mask != key:
+            return None
+
+        # Check that the key doesn't cover other keys that it shouldn't
+        next_key = keys[-1][0] + (2 ** mask_size)
+        max_key = key + (2 ** (mask_size + n_extra_mask_bits))
+        n_unused = max_key - (next_key & mask)
+        if n_unused > 0 and key_space_tracker.is_allocated(next_key, n_unused):
+            return None
+
+        return _AppKeyInfo(key, new_mask, core_mask, mask_size,
+                           keys[0][1].n_atoms * n_stages)
 
-        # Get the size of the matrices that will be required
+    @staticmethod
+    def __check_key_slices(n_atoms, slices):
+        # Check that the slices cover all atoms
+        slices = sorted(slices, key=lambda s: s.lo_atom)
+        slice_atoms = slices[-1].hi_atom - slices[0].lo_atom + 1
+        if slice_atoms != n_atoms:
+            return False
+        next_high = 0
+        for s in slices:
+            if s.lo_atom != next_high:
+                return False
+            next_high = s.hi_atom + 1
+        return True
+
+    def __app_key_and_mask(self, graph_mapper, m_edges, app_edge, routing_info,
+                           key_space_tracker):
+        # Work out if the keys allow the machine vertices to be merged
+        mask = None
+        keys = list()
+
+        # Can be merged only of all the masks are the same
+        pre_slices = list()
+        for m_edge in m_edges:
+            rinfo = routing_info.get_routing_info_for_edge(m_edge)
+            vertex_slice = graph_mapper.get_slice(m_edge.pre_vertex)
+            pre_slices.append(vertex_slice)
+            if rinfo is None:
+                return None
+            if mask is not None and rinfo.first_mask != mask:
+                return None
+            mask = rinfo.first_mask
+            keys.append((rinfo.first_key, vertex_slice))
+
+        if mask is None:
+            return None
+
+        if not self.__check_key_slices(
```

Comment:
i would wrap this and the checks that are inside __get_app_key_and_mask into its own method called can _do_app_pop_entry which then gives a 1 location for all the assumptions and checks. 

this checker needs to ensure n neurons is small enough to fit in the 11 bits given to the variable in the master pop entry.

Reply:
The check is done, but the method is left split up to make less confusing!